Create PostPairsTrading.md (#11)
hablapps/pairstrading: PostPairsTrading.md

> # A Match Made in Trading: Step-by-Step Pairs Trading Guide

Reviewer: Are you ok with this title?

This statistical test is a **hypothesis test**, where we use our data to see if we can accept or reject a hypothesis. In our case, the hypothesis is whether the time series is non-stationary. To determine this, we use **p-values**.

**P-values** help us decide whether to reject the null hypothesis. If the p-value is low, it indicates that we can reject the hypothesis that the time series is non-stationary, suggesting that our assets are cointegrated. The lower the p-value, the greater the confidence in rejecting the null hypothesis. It is very common to use a threshold of 0.05 on the p-value to reject hypotheses.

For the sake of simplicity, we will be using [PyKX](https://code.kx.com/pykx/2.4/index.html). This is necessary as we require importing our ADF test function and plotting a heatmap of our results. Developing these functionalities directly in Q might introduce errors and would be time-consuming, to say the least. Hence, we rely on PyKX to streamline the process by importing relevant libraries from the Python ecosystem.

```q
system "l pykx.q"
```

> Next, we need to import the **statsmodels library**, a prominent tool in Python for statistical modeling and hypothesis testing. It equips analysts with a robust toolkit for regression, time series, and multivariate analysis. Specifically, within the **statsmodels** package, the **statsmodels.tsa.stattools** module features **the Augmented Dickey-Fuller (ADF) test**.

Reviewer: "One of such libraries is **statsmodels**, a prominent tool..."

In this heatmap, they exhibit a vibrant green color, indicative of a high degree of cointegration, or, in simpler terms, a very low probability of not being cointegrated. They demonstrate low p-values suggesting their strength as candidates.

 > 💡 As we can see, this pair of indexes is not the best candidate according to our ADF tests. However, we chose it because the tick data for their prices is publicly available. We used TickStory to obtain the data.

![Prices](resources/cointegration.png)

The plotted graph displays the prices of both indexes together, providing a clearer comparison that showcases the cointegration between them. The blue line represents SP500, and NASDAQ100 is represented by the green line. The close alignment of their price movements indicates that they are cointegrated to some extent. This means that, despite short-term deviations, the indices tend to move together as time goes on, maintaining a stable relationship.

> In this case, we are using a [KX Dashboard](https://code.kx.com/dashboards/) to plot our data. We stream this data in one process to our local dashboard, which listens to that process and accesses the data to render visualizations.

Reviewer: I feel we are overexplaining things here, what about?
> This graph was generated by KX Dashboard, which receives data from the Pairs Trading process and renders visualizations.

Also, I think we could put it at the end of the previous paragraph.

By the way, the explanation is way far better with the brand new graph.

```q
q)log priceX
1.609438 2.302585 1.94591 1.386294 2.079442
q)log priceY
3.135494 3.401197 3.218876 3.401197 3.555348
q)spreads: log[priceY] - log priceX
1.526056 1.098612 1.272966 2.014903 1.475907
```

We're making progress, as we observe **numbers now fluctuating within much smaller ranges**. However, we're still missing a clear understanding of the underlying relationship. While we've normalized the data using logarithms, we now need to align their discrepancies to a single asset. 

Since both assets are related, **we can leverage linear regression** to our advantage. This enables us to simplify our spreads effectively. So, we'll conduct a basic linear regression analysis using historical data to discern the disparity between them. The generic formulae for one is:

$$Y = \alpha + \beta X + \varepsilon$$

In this context, Y represents the NASDAQ 100 index, X represents the S&P 500 index, α is the intercept, β is the slope (which indicates the relationship strength between the two indices), and ε is the error term.

> ![LinearRegression](resources/linear_regression.png)

Reviewer: I miss the regression line generated by the optimal alpha and beta.

Reviewer 2: done

The aim of this post is not to delve deeply into these methods but to mention that the most popular method is called the least squares method. For this case, it provides a closed-form solution that depends on our historical data. This means we do not need any iterative algorithm or a more complex method to find these optimal alpha and beta values.

>💡 For those interested in our implementation of these formulas in kdb+/q, the code can be found in our repository [Pair-Trading](https://github.com/hablapps/pairstrading/blob/5-Post/linear_regression.q).

This precisely meets our objective—a **comprehensive method for representing relative changes between both assets**. As we can deduce, our mean is now 0 because our assets are normalized, cointegrated and on the same scale. Therefore, ideally, the differential between their prices should be 0. Consequently, when our spread is below 0, we infer that asset X is overpriced, whereas if it's above 0, then asset Y is overpriced.

## Real-Time Pair Party

Now that we have selected a pair of cointegrated indices and understand how to calculate their relationships, let's see how we can create a real-time pair trading scenario.

> > ⚠️ An important note is that this post will include a real-time simulation. In other words, if we wanted to develop a 100% real-time product, we would need to make slight adjustments to the code.

Reviewer: Does this post include a real-time simulation? I think we aren't discussing that further on, I mean, we aren't showing data here. Maybe you just wanted to refer to the repo.

Reviewer 2: I think its more clear now

Reviewer: Could we move this warning further on? Later, you mention again that this is a simulation ("In our case, we are simulating real time; we do not have a 100% real-time product.") and I feel it's redundant. Could we integrate both references in the same paragraph?

As previously mentioned, historical data is crucial for generating accurate spreads. We need to calculate each spread in real-time **using precomputed alpha and beta values** derived from both prices. Therefore, once we have obtained our historical values, we can proceed with calculating our spreads from linear regression.

```q
spread: priceY[.streamPair.i][`bid] - ((priceX[.streamPair.i][`bid] * beta_lr)+alpha_lr);
```

> > 💡 You may notice that we retrieve bid price data from our price stream using an index (`.streamPair.i`). This occurs because we simulate the arrival of these records dynamically, based on a delta time, and thus read from our real-time simulated table, utilizing our updated index `.streamPair.i` with each new record.

Reviewer: I think I understand what is going on here, but it's hard to follow.

Reviewer 2: refactored

Using this approach, we will end up with something like this:

> ![SpreadsD](resources/spreads.gif)

Reviewer: This isn't loading properly in Github.

Reviewer: I still can't see this, so can't review.

And there we have it! **A perfectly plotted spread series in real-time**, ready to be utilized for further analysis and exploitation.

## What's left to start making a profit?

> Finally, once we have our spreads accurately calculated and observe how our data is being updated we can **execute buy and sell orders when spread discrepancies occur** based on some signal windows.

Reviewer: Could we provide a visual taste of it?

> 💡 Signal windows play a pivotal role in implementing Pairs Trading strategies. They serve as indicators for determining when to execute buy and sell actions, acting as arbitrary thresholds that guide our algorithm's decision-making process. These windows are derived from the variance of our data, representing a static variance assumption due to our consideration of a time-independent cointegrated series. However, we'll dig deeper into this topic in a subsequent post that will expand the scope of the current discussion.

After all this, you should be able to understand **our spread formulation and what it represents**. Basically, we can identify instances where one asset is overpriced while the other is underpriced.

> One might argue that our calculations are heavily influenced by past data and that we rely too much on historical changes that **may not accurately reflect the present reality**. This is indeed a **valid concern**. To address this issue, we could implement **a rolling window approach** where the linear regression is continuously updated. 

Reviewer: I think that we start here with future work, I'd move it to the conclusions.

Reviewer 2: refactored

> This would ensure that our model remains responsive to changes in the underlying data over time. Additionally, we can use the Kalman Filter to dynamically fit the alpha and beta of the linear regression. The Kalman Filter effectively filters noise and predicts states in a dynamic system, allowing for real-time adjustments and providing a more accurate reflection of the present market conditions.

Reviewer: The first sentence should go with the previous paragraph.

1. They have a similar trend, meaning the difference between both assets maintains a constant mean, and this difference fluctuates around that same mean.

2. This inherent relationship persists in the long run, meaning that our series is not dependent on time.

## A pair in the hand is worth two in the bush.

### ADF testing

Imagine **we selected 13 world indexes** and aimed to assess whether they are **cointegrated or not**. In this scenario, a crucial tool at our disposal is the Augmented Dickey-Fuller (ADF) test.

> 💡 Please note that for simplicity in this code, we will be using pykx. This is necessary as we need to import our ADF test function and plot a heatmap of our results.
 
> `system "l pykx.q"`

Next, we import the statsmodels.tsa.stattools library and define a custom cointegration function that returns a dictionary with the ADF test results given 2 assets. **We then apply our cointegration function to the Cartesian product of every pair of assets** to get a matrix with every p-value. 

![cointegration](https://github.com/hablapps/pairstrading/blob/5-Post/CointAlGif.gif?raw=true)

---

```q
// Recieves 2 symbols and returns a dictionary

> fCoint {[assetX;assetY;]

Reviewer 2: Show more q code. Maybe show how you can make use of Python code using PyKx

Author: Done, Altough I did not add examples of pykx use (just the line of code where we import our `coint` function). So I will not resolve this conversation because maybe you want me to add some more examples

        ... }

> crossedList: symList cross symList

Reviewer 2: explain that `cross` is a q function to generate the cartesian product

> matrix: fCoint .' crossedList;

Reviewer 2: something similar here, explain that this is an apply with an each, and add a reference to the doc

## Spreading spreads

Indeed, just subtracting the prices of two assets, as in: $priceY−priceX$ may not provide a clear understanding of their relationship. Let's illustrate this with an example:

Consider the following series:

---

```q
priceX: 5 10 7 4 8

priceY: 23 30 25 30 35

spreads: priceY - priceX // = 18 20 18 26 27
```

> 💡 As you can see and verify through Q/kdb+ array properties, spreads can be calculated by simply computing the difference between two vectors without the need for special functions or loops.

---

〽 **These spread values don't offer much insight** into the relationship between the two assets. Are both assets increasing? Are they moving in opposite directions? It's unclear from these numbers alone.

Let's consider **using logarithms**, as they possess favorable properties for our pricing model. They inherently prevent negative values and tend to approach zero:

--- 

```q
log priceX = 1.609438 2.302585 1.94591 1.386294 2.079442

log priceY = 3.135494 3.401197 3.218876 3.401197 3.555348

> spreads: log priceY - log priceX // = 1.526056 1.098612 1.272966 2.014903 1.475907

Reviewer 2: there is a bug here, you need the use of parentheses or brackets
```suggestion
spreads: log[priceY] - log priceX // = 1.526056 1.098612 1.272966 2.014903 1.475907

```
And i think is a good moment to explain that q has no more preference order than right to left

```

---

We're making progress, as we observe **numbers now fluctuating within much smaller ranges**. However, we're still missing a clear understanding of the underlying relationship. While we've normalized the data using logarithms, we now need to align their discrepancies to a single asset. 

Since both assets are related, **we can leverage linear regression** to our advantage. This enables us to simplify our spreads effectively. So, we'll conduct a basic linear regression analysis using historical data to discern the disparity between them:

--- 

$spread = log(priceY) - (beta * log(priceX)+alpha)$

```q
historical_data_priceX: 7 10 6 5 8

historical_data_priceY: 23 25 16 20 15

beta: 0.2679227

alpha: 2.444817
```

---

We've already calculated the alpha and beta using the logarithmic values of our historical data (since we don't have prior knowledge of the real-time price values for priceX and priceY). Now, all that remains is to apply the previous formula to derive our spreads:

---

> $spreads$ = -0.1493929 0.0451223 -0.08835117 0.0451223 0.1579725

Reviewer 2: I think this is quite hard to understand (there are so much decimals), maybe you dont need to wirte the entire formula, maybe you can just say that just evaluating you get that result

![Spreads](https://github.com/hablapps/pairstrading/blob/5-Post/Spreads.gif?raw=true)

## Two steps forward, one step back.

Before proceeding to plot the NASDAQ100-SP500 spreads, we need to first plan our algorithm. In this post, we intend to create **a real-time scenario** for Pairs trading, so careful planning is essential for our sake.

Decomposing our steps, let's start from the very beginning:

1. Data management part

        1.1. Read data

        1.2. Filter data

2. Outside the loop

        2.1. Linear regression

        2.2. Initialize values 

3. Inside the loop

        3.1. Calculate spreads

        3.2. Write on buffers

As previously mentioned, historical data is crucial for generating accurate spreads. We need to calculate each spread in real-time **using precomputed alpha and beta values** derived from both prices. Therefore, once we have obtained our historical values, we can proceed with calculating our linear regression.

---

```q
// Fix data and take log(prices) -> Simulated data

> priceX: 0!1_(update delta:0f^deltas dateTime from distinct select distinct dateTime, log bid, log ask from (update dateTime:"P"$@[;19;:;"."] each dateTime from tab1) where not null bid);

Reviewer 2: this is a long line, divide this into maybe 3 different lines?

Reviewer 2: and also I think is a good moment to explain briefly qsql. Maybe you can say that q is quite hard to learn, but qsql is a vey good point to start

Author: I did explained QSQL but not in this line but before

> priceY: 0!1_(update delta:0f^deltas dateTime from distinct select distinct dateTime, log bid, log ask from (update dateTime:"P"$@[;19;:;"."] each dateTime from tab2) where not null bid);

Reviewer 2: again here

// Read historical data
historial_tab2: 1_ flip `open`high`low`close`adjClose`vol!("FFFFFF";",") 0: `:/data/stocks/NASDAQ100_hist.csv;
historial_tab1: 1_ flip `open`high`low`close`adjClose`vol!("FFFFFF";",") 0: `:data/stocks/SP500_hist.csv;
```

---

## Linear regression

Now, armed with logarithms, we can replicate the process from our previous example and calculate both alpha and beta values. To do this, we'll start by taking the historical data for NASDAQ100 and SP500 and applying our linear regression function to derive the desired values.

---

```q
// Calculate alpha and beta from historical values

> beta_lr: betaF[px:-100#log historial_tab1`close;py:-100#log historial_tab2`close]; // we only take most recent 100 values 

Reviewer 2: explain that `betaF` and `alphaF`are the functions to fit linear regression

alpha_lr: alphaF[px;py];
```

---

Now we just calculate our spreads as we did before like:

---

```q
spread: priceY[.streamPair.i][`bid] - ((priceX[.streamPair.i][`bid] * beta_lr)+alpha_lr);
```

---

> 💡 You may notice that we retrieve bid price data from our price stream using an index (`.streamPair.i`). This is because we are simulating the arrival of these records dynamically, based on some delta time.

This approach will provide us with:

> ![SpreadsD](https://github.com/hablapps/pairstrading/blob/5-Post/SpreadsD.JPG?raw=true)

Reviewer 2: Make a gif of this :) . And also could you show the part of the dashboard where you just plot the prices?

Author: I will tomorrow or so :D

Recapping, we have covered:

1. An introductory overview of the market as a whole.
2. An examination of cointegrated assets within the market.
3. Multiple Augmented Dickey-Fuller (ADF) tests on real assets.
4. A presentation of the pairs trading strategy itself.
5. A clear and guided explanation of spreads calculation, interpretation, and implementation in Q/kdb+.
6. Additional knowledge necessary to master the strategy.

We aimed to demonstrate the capabilities of Q/kdb+ and its potential in a simplified manner that anyone can implement, particularly in the context of a widely used financial strategy. By doing so, we hope to make complex concepts more accessible and empower individuals to leverage these powerful tools in their own endeavors.

We hope you found this information valuable and gained a good understanding of this financial tactic from both technical and economical perspectives. If you have any questions or need further clarification, don't hesitate to reach out. 

Be sure to stay tuned for more posts and updates on this blog to deepen your knowledge even further

> Special thanks to [...] for [...]

Reviewer 2: Add references pls

Linear regression aims to identify relationships between historical data, which we then extrapolate to current data. The differences between these relationships, or deviations, are our spreads. We've already calculated the 𝛼 and 𝛽 using the logarithmic values of our historical data (since real-time price values for priceX and priceY are unknown). Now, we simply combine everything and apply linear regression to our price logarithms:

$$spread = log(priceY) - (\beta \cdot log(priceX)+\alpha)$$

```q
q)spreads: log[priceY] - alpha + log[priceX] * beta
-0.1493929 0.0451223 -0.08835117 0.0451223 0.1579725
```

There are different methods we can use to obtain the best alpha and beta values that minimize the spreads or, in other words, there are mathematical methods to find the line that best fits the prices.

> The aim of this post is not to delve deeply into these methods but to mention that the most popular method is called the least squares method. For this case, it provides a closed-form solution that depends on our historical data. This means we do not need any iterative algorithm or a more complex method to find these optimal alpha and beta values.

Reviewer: Six occurrences of _method_ in just two paragraphs, could we reduce them?

This precisely meets our objective—a **comprehensive method for representing relative changes between both assets**. As we can deduce, our mean is now 0 because our assets are normalized, cointegrated and on the same scale. Therefore, ideally, the differential between their prices should be 0. Consequently, when our spread is below 0, we infer that asset X is overpriced, whereas if it's above 0, then asset Y is overpriced.

## Real-Time Pair Party

Now that we have selected a pair of cointegrated indices and understand how to calculate their relationships, let's see how we can create a real-time pair trading scenario.

> ⚠️ An important note is that this post will include a real-time simulation. In other words, if we wanted to develop a 100% real-time product, we would need to make slight adjustments to the code.

The first step to implementing this pair trading algorithm in real time is to declare a `.z.ts` function. This `.z.ts` function will be called automatically every x milliseconds which can be configured with `\t`. In our case, it will be called every 100 milliseconds.

```q
.z.ts: {.streamPair.genPair[]} 
\t 100
```

> Let's now see how our **.streamPair.genPair** function should be defined. In our case, we are simulating real time; we do not have a 100% real-time product. Therefore, we already have the data loaded into memory and only need to display it one by one. For this, we will use an index **streamPair.i** which we will update with each execution of our function. 

Reviewer: Use a different connector: "In our case" was used in the previous paragraph as well.

```q
.streamPair.i+:1;
resX: priceX[.streamPair.i];
resY: priceY[.streamPair.i];
```

The purpose of this function is to calculate the corresponding price spreads. For this, we will use the spread formula that we already know.

```q

> spread: priceY[.streamPair.i][`bid] - ((priceX[.streamPair.i][`bid] * beta_lr)+alpha_lr);

Reviewer: Why not using `resX` and `resY` here as well? I miss this expression to be exactly the same as the one you show later on in `.streamPair.genPair`.

Reviewer 2: true, changed

```

Putting everything together and returning a table with the time instant and the spread, we would get the function:

```q
 .streamPair.genPair:{
      .streamPair.i+:1;
      resX: priceX[.streamPair.i];
      resY: priceY[.streamPair.i];
      s: resY[`bid] - alpha_lr+resX[`bid] * beta_lr;
      enlist `dateTime`spread`mean!

>             ("p"$(resX[`dateTime]);"f"$(s);"f"$(0));  

Reviewer: So `mean` is always 0? Do we need parentheses?

I think it's ok to have a simplification but I don't get the objective of this function, I mean, it just returns a dictionary but nobody is collecting the result into a table or sending it to the dashboard. Maybe I'm getting it wrong.

Reviewer 2: Yes, when i simplificated i missed the part where we publish. I've just added it :)

 }
```

Using this approach, we will end up with something like this:

![SpreadsD](resources/spreads.gif)

And there we have it! **A perfectly plotted spread series in real-time**, ready to be utilized for further analysis and exploitation.

## What's left to start making a profit?

Finally, once we have our spreads accurately calculated and observe how our data is being updated we can **execute buy and sell orders when spread discrepancies occur** based on some signal windows.

A simple approach to window signals is to set these windows as twice the historical standard deviation of the spreads. Therefore, if either of these limits is reached, we should sell the overvalued index and buy the undervalued one, and then unwind our position when the spread returns to 0. Let's clarify this with a specific example:

> ![SpreadsD](resources/window_signals.gif)

Reviewer: This is very nice! 👍 

Just one minor thing, what are the values in the X axis?

Reviewer 2: The x-axis should represent time, but we struggled to plot in real time because it froze when we tried to edit the x-axis labels.

# A Match Made in Trading: Step-by-Step Pairs Trading Guide

KDB+/Q stands out as **a powerful tool in finance**, renowned for its ability to handle vast volumes of real-time data amidst the relentless dynamics of the market. In this article, we embark on an insightful exploration of Pairs Trading and its implementation in Q, offering a comprehensive guide to one of the most popular strategies in the trading world.

> Our objective is to **provide an easy-to-understand explanation of some of the intricacies of pair trading**, bridging the gap between theory and practice.

Reviewer: This is a derived objective, but not the main one. With the new perspective, the main objective is:
* Implementing Pairs Trading strategy in q
  + Identify cointegrated indexes
  + Implement model to calculate spreads
* Get it working in real-time
  + Integrate with tick architecture

We'll proceed methodically, ensuring each question leads to a comprehensive answer. To start, we'll contextualize our current situation by addressing key questions such as **"What do we know about the market and how can we benefit from it?"**. This will lay the foundation for constructing a real-time simulated environment on Pairs Trading that will exemplify everything we have explained thus far.

Whether taking a technical or quantitative approach, these insights will provide valuable foundations for constructing this algorithm effectively.

We will be covering the following aspects using the KDB tick architecture as a base:

![Architecture](resources/general-architecture.png)

> As you can see, there are a few new components added to the picture. The details will be discussed as we go on, but I'm sure you can already recognize some familiar faces.

Reviewer: I think we should assume that this is their very first contact with kdb (so they shouldn't recognise elements from the diagram). Here I would clarify that this article is aimed at quants in order to motivate them to adopt kdb+/q, due to its incredible performance and concision. But also, we'll supply a brief introduction to Pairs Trading so other developers can benefit from it as well.

Is this described mathematically? **Yes**:

The concept we're referring to is **cointegration** (although there are other methods, we'll focus on this one).

> 💡 Which should not be confused with correlation; cointegration is a statistical property of two-time series, indicating a long-term relationship between them despite short-term fluctuations. Cointegrated series move together over time, sharing a common stochastic drift. On the other hand, correlation measures the strength and direction of the linear relationship between two variables at a specific point in time. While correlation captures the degree of association between variables, cointegration reflects a deeper, underlying relationship that persists over time.

Hence, we're interested in **cointegrated assets**, which are assets that exhibit the following characteristics:

1. They have a similar trend, meaning the difference between both assets maintains a constant mean, and this difference fluctuates around that same mean.

> 2. This inherent relationship persists in the long run, meaning that our series is not dependent on time.

Reviewer: I think that we should add a new section, which supplies a very brief intro to the components of the classical Tick Architecture. I would end up clarifying the next section. First, we'll identify cointegrated pairs. Then, we'll produce the Pairs Trading model to calculate spreads. Finally, we'll exploit such model from a real-time component and will visualise its outcomes.

## A pair in the hand is worth two in the bush

> ### ADF testing

Reviewer: Now, this section becomes "Identifying cointegrated indexes" or something like that...

> Imagine **we selected 13 world indexes** and aimed to assess whether they are **cointegrated or not**. In this scenario, a crucial tool at our disposal is the Augmented Dickey-Fuller (ADF) test, an essential statistical test for assessing the stationarity of time series data. The more stationary the time series are, the more cointegrated they are likely to be.

Reviewer: We could introduce `sym`s right after the first sentence.

This statistical test is a **hypothesis test**, where we use our data to see if we can accept or reject a hypothesis. In our case, the hypothesis is whether the time series is non-stationary. To determine this, we use **p-values**.

**P-values** help us decide whether to reject the null hypothesis. If the p-value is low, it indicates that we can reject the hypothesis that the time series is non-stationary, suggesting that our assets are cointegrated. The lower the p-value, the greater the confidence in rejecting the null hypothesis. It is very common to use a threshold of 0.05 on the p-value to reject hypotheses.

> For the sake of simplicity, we will be using [PyKX](https://code.kx.com/pykx/2.4/index.html). This is necessary as we require importing our ADF test function and plotting a heatmap of our results. Developing these functionalities directly in Q would be time-consuming and prone to errors. Although an implementation of the ADF test in KDB+/Q would be more efficient and faster, the effort required would outweigh the benefits. Therefore, we rely on PyKX to streamline the process by leveraging relevant libraries from the Python ecosystem.

Reviewer: ... outweigh the benefits **for this particular case** where performance isn't critical.

Reviewer: Why don't we import pykx right here as we did before?

> One such library is **statsmodels**, a prominent tool in Python for statistical modeling and hypothesis testing. It equips analysts with a robust toolkit for regression, time series, and multivariate analysis. Specifically, within the **statsmodels** package, the **statsmodels.tsa.stattools** module features **the Augmented Dickey-Fuller (ADF) test**.

Reviewer: Same here, why don't we import statsmodel at this point?

> As we saw in the introduction, we are working in a tick architecture environment. This means that, in addition to receiving real-time prices for our indices, this architecture provides a tool to store the closing prices of our indices in our historical database (HDB) at the end of the day.

Reviewer: Don't mention real-time here. We are just dealing with the HDB that we introduced before.

Therefore, we simply need to execute a straightforward query on the HDB to read these data and load them into memory. To achieve this, we define the function `rs`, which takes a date range and the indices for which we want to retrieve data. We can then use the **qSQL syntax** (very similar to SQL) to obtain the desired data.

```q
rs:{[id;ed;syms]select from prices where date within (id;ed),sym in syms}
```

Now we simply need to send this function with the necessary parameters to our HDB. To do this, we open (`hopen`) a connection to our HDB process, obtaining a handle. To communicate with the process, we pass a list to the handle with the first element being the function and the subsequent elements being the parameters, once we get our data we finally group (`xgroup`) by index. Finally, let's not forget to close (`hclose`) the connection to HDB.

```q
syms:`SP500`NASDAQ100`BFX`FCHI`GDAXI`HSI`KS11`MXX`N100`N225`NYA`RUT`STOXX
ed:2024.03.30    / end_date
tr:4*365         / date range     
h:hopen port

> t:`sym xgroup h(rs;ed-trange;ed;syms)

Reviewer: `trange` is not defined, I think it's `tr` instead.

Reviewer: Maybe we could simplify things by assuming that the end date is today (`.z.d`) and just supplying the `n` previous days.

Also, what do you think of renaming `h` to `hdb`? Not sure what to do with "port", maybe we could just indicate that `h` (or `hdb`) is a connection with the HDB process and just include a reference to q for mortals on that for more detail.

hclose h
```

> We then proceed to create a function called **fcoint** to call our imported function from PyKX, handle any null values by filling them with 0, using `0f^` and return the second value, which in this case is the p-value.

Reviewer: I think that we could introduce both before. We could generate the cartesian products at the very end, since we are about to evaluate the ADF test against all pairs. `fcoint` could be introduced along with `statsmodel`.

```q
system"l pykx.q"
coint:.pykx.import[`statsmodels.tsa.stattools]`:coint
fcoint:{@[;1]0f^coint[x;y]`}
```

We generate all combinations (`cross`) of indexes to see which pair is most cointegrated. Then, we index (`@`) each pair in our table. Additionally, we take (`#`) the last **trange** days of data for both indexes, and finally apply our **fcoint** function to each (`.'`) pair of data lists.

```q
trange:4*252
matrix:fcoint .' 0f^@\:[;`close](@/:[t]')syms cross syms
```

Now, with our matrix in hand, we can plot it and **visually identify** which asset is more favorable. In order to do that, we can leverage PyKX once again to bring the `heatmap` module to q:

> 💡 We could have created a dashboard to plot the heatmap using KX Dashboard, but in this case, it is simpler and faster to use PyKX and plot as we would in Python, with minor modifications to the syntax.

```q

> pyhm:.pykx.import[`seaborn]`:heatmap

Reviewer: Could we briefly introduce seaborn in the previous sentence?

```q
pyshow:.pykx.import[`matplotlib.pyplot]`:show
pyshow[::]
```

Given our following assets:

| SP500 | NASDAQ100 |   BFX   |  FCHI  |  GDAXI  |    HSI    |  KS11   |  MXX   |  N100  | N225  |  NYA  |  RUT  | STOXX  |
| :---: | :-------: | :-----: | :----: | :-----: | :-------: | :-----: | :----: | :----: | :---: | :---: | :---: | :----: |
|  USA  |    USA    | Belgium | France | Germany | Hong Kong | S.Korea | Mexico | Europe | Japan |  USA  |  USA  | Europe |

Our heatmap looks like this:

![ADF heatmap](https://github.com/hablapps/pairstrading/blob/5-Post/resources/ADFgif.gif?raw=true)

As we can observe, there are several cointegrated indices, but our attention will be drawn towards the **NASDAQ100 and SP500** synergy. Both of these indices belong to the American market and share numerous characteristics. They encompass American companies traded within the same scenario, which is what makes them a perfect fit for our case.
In this heatmap, they exhibit a vibrant green colour, indicative of a high degree of cointegration, or, in simpler terms, a very low probability of not being cointegrated. They demonstrate low p-values suggesting their strength as candidates.

>  > 💡 As we can see, this pair of indexes is not the best candidate according to our ADF tests. However, we chose it because the tick data for their prices is publicly available. We used TickStory to obtain the data.

Reviewer: How is it possible that we could calculate the cointegration for all indexes, but we can only access NASDAQ100 and SP500 because their data is publicly available? I'm missing something here...

![Prices](resources/cointegration.png)

> The plotted graph displays the prices of both indexes together, providing a clearer comparison that showcases the cointegration between them. The blue line represents SP500, and NASDAQ100 is represented by the green line. The close alignment of their price movements indicates that they are cointegrated to some extent. This means that, despite short-term deviations, the indices tend to move together as time goes on, maintaining a stable relationship. This graph was generated by [KX Dashboard](https://code.kx.com/dashboards/), which receives data from the Pairs Trading process and renders visualizations.

Reviewer: This is a good moment to show the tick architecture diagram and recap that we have implemented the ADF box and that its output would be helpful for the next section. Alternatively, we could introduce it at the very beginning of this section to guide the explanations, whatever you prefer!

## Cointegration, then what?

Let's recap our progress:

1. We've acknowledged the randomness of the market, yet discovered that certain assets **exhibit similar movements**.

2. By employing the **cointegration method and the ADF test**, we pinpointed a promising pair of assets for our analysis: NASDAQ100 and SP500.

Now we're faced with a crucial question: **"How can I benefit from this knowledge?"**

As mentioned earlier, the market is inherently random and doesn't always behave predictably. While NASDAQ100 and SP500 often follow similar trends, their individual values **can sometimes diverge significantly**. For instance, NASDAQ100 may rise while SP500 falls, or vice versa. 

> However, this presents **an opportunity for profit** because we know that these assets tend to revert to their shared mean over time. If one asset is **overpriced** and likely to decrease, we may consider **selling it** (going short). Conversely, if an asset is **underpriced** and expected to increase, we may consider **buying it** (going long). That is what we call Pairs Trading.

Reviewer: We should move this lines at the end of the previous section. This is exactly the kind of summary that should rely on the tick architecture diagram.

> 💡 This strategy possesses financial characteristics: our **profitability remains unaffected by the broader market trends**, as our focus lies solely on the disparity between the two assets. It's about relative movements rather than absolute ones; we're indifferent to whether prices are rising or falling. This quality defines it as a **neutral market strategy**.

> ## Spreading spreads

Reviewer: Ok, now let's go for the Pair Trading model that calculates spreads. Could we find a nicer name for this section?

The most common method to find the best relationships (alpha and beta) is the least squares method, which minimizes the sum of the squared residuals:
$$S(\alpha, \beta) = (log(priceY) - (\beta \cdot log(priceX)+\alpha)^2$$

After taking partial derivatives with respect beta and setting to zero, and then solving, we can arrive at this formula:

$$\beta = \frac{{(n \cdot \sum(x \cdot y)) - (\sum x \cdot \sum y)}}{{(n \cdot \sum(x^2)) - (\sum x)^2}}$$

Which we can see implemented in the following functions:

```q
betaF:{dot:{sum x*y};                                      
      ((n*dot[x;y])-(*/)(sum')(x;y))%                         
      ((n:count[x])*dot[x;x])-sum[x]xexp 2}
```

Now, following the same steps as before but for alpha, we arrive at:

$$\alpha = \bar y - \beta \cdot \bar x$$

Which is implemented in the following line of code:

```q
alphaF: {avg[y]-betaF[x;y]*avg[x]}
```

Now we simply need to apply these functions to find the optimal alpha and beta on the historical prices (which we took from HDB) of the indices we choose.

```q
beta_lr:betaF[t[`SP500]`close;t[`NASDAQ100]`close]
alpha_lr:alphaF[t[`SP500]`close;t[`NASDAQ100]`close]
```

> This precisely meets one of our objectives: getting **a comprehensive method for representing relative changes between both assets**. As we can deduce, our mean is now 0 because our assets are normalized, cointegrated and on the same scale. Therefore, ideally, the differential between their prices should be 0. Consequently, when our spread is below 0, we infer that asset X is overpriced, whereas if it's above 0, then asset Y is overpriced.

Reviewer: Again, we should rely on the diagram here. We can emphasise the HDB, the line that reads from the HDB and the RPT (but not emphasising dashboards).

To do this, we need to focus on the Real-Time Pair Trading (RPT), which will subscribe to the tickerplant, receiving the price data (`prices`) of the two indices we choose based on our ADF Test. Additionally, we will connect to the HDB (as shown in the ADF testing) to calculate the alpha (`alpha_lr`) and beta (`beta_lr`) of the linear regression. Subsequently, the RPT will publish the regression spreads to the KX Dashboard so we can view it in real time. Let's take a quick look at the RPT script.

The first step is to declare a `.z.ts` function, which will be called automatically every x milliseconds, configurable with `\t`. In our case, it will be called every 100 milliseconds. This function will publish the spreads in real time to a table using the `.u.pub` (publish) function from the [KDB+ tick architecture](https://github.com/KxSystems/kdb-tick). The `.u.pub` function takes two parameters: the name of the table to publish to and the content to be published, then it publishes the content to the table's subscribers.

```q
.z.ts: {.u.pub[`spreads;update priceY - alpha_lr+priceX * beta_lr from prices]} 
\t 100

> ```

Reviewer: I miss the `pub` method from the RPT component itself, where you accumulate on the `prices` table. Which component is responsible for cleaning the published spreads?

By using this approach, we only need to connect KX Dashboards to our publisher by setting up a new connection from the connection selector in the UI.
This will allow us to plot our spreads in real time and we will end up with something like this:

![SpreadsD](resources/spreads.gif)

And there we have it! **A perfectly plotted spread series in real-time**, ready to be utilized for further analysis and exploitation.

> ## What's left to start making a profit?

Reviewer: I'm not sure if we should create another section for this, maybe we could place this at the end of the previous one.

![WSignals](resources/window_signals.gif)

In this instance, we can see that the spread (purple line) is positive and above the signal (blue line), indicating that our Y index (NASDAQ100) is overvalued relative to the SP500. Therefore, we should sell NASDAQ100 and buy SP500. At the end of the gif, it can be observed that the spread returns to 0 (green line), meaning the indexes are no longer overvalued or undervalued, respectively. At this point, we should unwind the positions we acquired earlier.

> 💡 Signal windows play a pivotal role in implementing Pairs Trading strategies. They serve as indicators for determining when to execute buy and sell actions, acting as arbitrary thresholds that guide our algorithm's decision-making process. These windows are derived from the variance of our data, representing a static variance assumption due to our consideration of a time-independent cointegrated series.

> ## Conclusion and Future Work

Reviewer: The content is fine, but requires some reordering. We should explicitly address the objectives that we stated in the introduction. Future work is fine.

Our objective is to provide **an implementation of a Pair Trading strategy in q**, which includes identifying cointegrated indexes, implementing a model to calculate spreads, making it work in real-time, and integrating it with the tick architecture.

We'll proceed methodically, ensuring each question leads to a comprehensive answer. To start, we'll contextualize our current situation by addressing key questions such as **"What do we know about the market and how can we benefit from it?"**. This will lay the foundation for constructing a real-time simulated environment on Pairs Trading that will exemplify everything we have explained thus far.

Whether taking a technical or quantitative approach, these insights will provide valuable foundations for constructing this algorithm effectively.

We will be covering the following aspects using the KDB tick architecture as a base:

> ![Architecture](resources/general-architecture.png)

Reviewer: Very important: we need to add a reference to the original picture from DefConQ

```q
t:`sym xgroup hdb(query_prices;trange;syms)
:{[tr;syms]select from prices where date within (.z.d-tr;.z.d),sym in syms}
```

Now we simply need to send this function with the necessary parameters to our HDB. To do this, we open (`hopen`) a connection to our HDB process, obtaining a handle. To communicate with the process, we pass a list to the handle with the first element being the function and the subsequent elements being the parameters, once we get our data we finally group (`xgroup`) by index. Finally, let's not forget to close (`hclose`) the connection to HDB.

```q
tr:4*365         / date range     
hdb:hopen port
t:`sym xgroup hdb(query_prices;tr;syms)
hclose h
```

 In our case we are going to use closing prices to our ADF test, so we have to index (`@`) by column **close** from each pair in our table. Additionally, we take (`#`) the last **tr** days of data for both indexes, and finally apply our **fcoint** function to each (`.'`) pair of data lists.

```q

> matrix:fcoint .' 0f^@\:[;`close](@/:[t]')syms cross syms

Reviewer: Would it make sense to just select `sym` and `close` in `rs`? (Take a look at my proposed text #13 to understand why)

```

Now, with our matrix in hand, we can plot it and **visually identify** which asset is more favorable. In order to do that, we can leverage PyKX once again to bring the `heatmap` module from `seaborn` a prominent data visualization library in the Python ecosystem to q:

> 💡 We could have created a dashboard to plot the heatmap using KX Dashboard, but in this case, it is simpler and faster to use PyKX and plot as we would in Python, with minor modifications to the syntax.

```q
pyhm:.pykx.import[`seaborn]`:heatmap

> pyhm[pvalues;`xticklabels pykw syms;`yticklabels pykw syms;`cmap pykw `RdYlGn_r]

Reviewer: I'd add a very short sentence describing what is going on here. I'm unable to decipher the `RdYlGn_r`.

![Arch 1st Part](resources/general-architecture-top.png)

As mentioned earlier, the market is inherently random and doesn't always behave predictably. While NASDAQ100 and SP500 often follow similar trends, their individual values **can sometimes diverge significantly**. For instance, NASDAQ100 may rise while SP500 falls, or vice versa. 

However, this presents **an opportunity for profit** because we know that these assets tend to revert to their shared mean over time. If one asset is **overpriced** and likely to decrease, we may consider **selling it** (going short). Conversely, if an asset is **underpriced** and expected to increase, we may consider **buying it** (going long). That is what we call Pairs Trading.

> 💡 This strategy possesses financial characteristics: our **profitability remains unaffected by the broader market trends**, as our focus lies solely on the disparity between the two assets. It's about relative movements rather than absolute ones; we're indifferent to whether prices are rising or falling. This quality defines it as a **neutral market strategy**.

> ## Determining how to calculate the spreads

Reviewer: I want to discuss two aspects from this section with you. Let me know when you find the time! Just as a reminder for myself:
* HDB -> RPT link
* price_y and price_x source (rs?)

The first thing we need to know is how tickerplant sends the data. What it does is to send a call to an `upd` function (that we must have declared in our process) with two arguments, the first one is the name of the table to which the incoming data belongs as the second argument. Then the first thing to do is to declare an `upd` function. What this function should do is to publish in a table (`spread`) the spreads of the linear regression. For this we will use the `u.pub` function of the tick architecture.

> 💻 The `.u.pub` function takes two parameters: the name of the table to publish to and the content to publish, then it publishes the content to the subscribers of the table, again calling the `upd` method that should be declared in the subscriber process.

```q
upd:{.u.pub[`spread;([]time:1#y`time;spread:sp . y`bid)]};
```

We simply need to declare an `sp` function that calculates the spread given the prices, but this is something we already know how to do.

```q

> sp:{y - alpha_lr + beta_lr * x};

Reviewer: This should be defined in the previous section, since the model is an input for this step. I prefer hiding alphas and betas in this section. What do you think?

# A Match Made in Trading: Step-by-Step Pairs Trading Guide

KDB+/Q stands out as **a powerful tool in finance**, renowned for its ability to handle vast volumes of real-time data amidst the relentless dynamics of the market. In this article, we embark on an insightful exploration of Pairs Trading and its implementation in Q, offering a comprehensive guide to one of the most popular strategies in the trading world.

> Our objective is to **provide a deep understanding of the intricacies of pair trading**, bridging the gap between theory and practice. Through a blend of theoretical insights and practical examples, we aim to equip you with the knowledge and skills necessary to navigate every aspect of this financial modelling strategy.

Reviewer: Maybe we are promising too much? 😄

Hence, we're interested in **cointegrated assets**, which are assets that exhibit the following characteristics:

1. They have a similar trend, meaning the difference between both assets maintains a constant mean, and this difference fluctuates around that same mean.

2. This inherent relationship persists in the long run, meaning that our series is not dependent on time.

## A pair in the hand is worth two in the bush

### ADF testing

Imagine **we selected 13 world indexes** and aimed to assess whether they are **cointegrated or not**. In this scenario, a crucial tool at our disposal is the Augmented Dickey-Fuller (ADF) test, an essential statistical test for assessing the stationarity of time series data. The more stationary the time series are, the more cointegrated they are likely to be.

> For the sake of simplicity, we will be using [PyKX](https://code.kx.com/pykx/2.4/index.html). This is necessary as we require importing our ADF test function and plotting a heatmap of our results. Developing these functionalities directly in Q might introduce errors and would be time-consuming, to say the least. Hence, we rely on PyKX to streamline the process by importing relevant libraries such as statsmodels.

Reviewer: Introduce statsmodel in the following paragraph.
"such as statsmodel" -> "from the Python ecosystem"

```q
system "l pykx.q"
```

> Next, we need to import the **statsmodels library**, a prominent tool in Python for statistical modeling and hypothesis testing. It equips analysts with a robust toolkit for data analysis, encompassing regression analysis, time series analysis, and multivariate analysis. Specifically, within the **statsmodels** package, the **statsmodels.tsa.stattools** module features **the Augmented Dickey-Fuller (ADF) test**.

Reviewer: Too many "analysis", would it be ok to say "robust toolkit for data, regression, time series and multivariate analysis"? Not sure if this change may affect the meaning though.

```q
coint:.pykx.import[`statsmodels.tsa.stattools]`:coint
```

> For our study, we retrieved data for the different indexes using the Yahoo Finance API and stored them in the `data/stocks/` directory. Additionally, for simplicity, we only use the closing prices (float), but the API also provides other typical values such as high, low, and open prices.

Reviewer: Supply link to Yahoo Finance API.

.../data/stocks directory, where we'll find one csv file for each index.

Is it open prices or opening prices? Previously you mention closing prices.

Reviewer 2: I think both "open prices" and "opening prices" are used in financial contexts

We declare the function **read_stock** to read the closing data of a given index.

> This function uses `0:` to read the files, which takes the delimiter and the schema. In this case, we only want to read the closing price column as a float. Additionally, since the data does not include any reference to the index being read, we need to make a small adjustment to our table to add the index associated with each price. Then, we apply this function to `each` of the indexes from which we want to read the data, and afterwards concatenate (`raze`) all the data into a table, and finally group (`xgroup`) by index. We declare a variable, **syms**, as a list of symbols, representing each of the indexes we want to check for cointegration.

Reviewer: "Then, we apply..., and afterwards..., and finally..." I think we can get rid of "and afterwards" to improve readability.

```q
read_stock:{[sym1]

>   update sym: sym1 from 1_ flip enlist[`close]!((5#" "),"F";",") 0:`$":data/stocks/",string[sym1],".csv"}

Reviewer: What about...?
```
rs:{([]sym:x;close:first((5#" "),"F";csv) 0:`$":data/stocks/",string[x],".csv")}
```

syms:`SP500_hist`NASDAQ100_hist`BFX`FCHI`GDAXI`HSI`KS11`MXX`N100`N225`NYA`RUT`STOXX

> superTab: `sym xgroup raze read_stock each syms

Reviewer: I would call this just `t`.

```

> We then proceed to create a function called **fCoint** to call our imported function from PyKX, handle any null values by filling them with 0, using `0f^` and return the second value, which in this case is the p-value.

Reviewer: A brief quote about _p-values_ could be nice.

Reviewer 2: done

```q
fCoint: {@[;1]0f^coint[0f^x;0f^y]`}
```

We generate all combinations (`cross`) of indexes to see which pair is most cointegrated. Then, we index (`@`) each pair in our table. Additionally, we take (`#`) the last **trange** days of data for both indexes, and finally apply our **fCoint** function to each (`.'`) pair of data lists. **trange** symbolizes the number of working days in the last 4 years.

```q
trange:4*252

> matrix: fCoint .' neg[trange]#''@\:[;`close](@/:[superTab]')syms cross syms

Reviewer: I think that you could move the `0f^` from `fCoint` here, right before `neg` and exploit ^ is atomic to invoke it just once.

Reviewer: This code is hard to follow, mostly because of the @/:' combination, but the explanation paves the way, so I guess it's ok.

Reviewer 3: First comment is done. Though I have not removed the first 0f^ from fCoint, since I don't know what the output of the coint function could be...

| USA   | USA       | Belgium | France | Germany | Hong Kong | S.Korea | Mexico | Europe | Japan | USA | USA | Europe |

Our heatmap looks like this:

![ADF heatmap](https://github.com/hablapps/pairstrading/blob/5-Post/resources/ADFgif.gif?raw=true)

As we can observe, there are several cointegrated indices, but our attention will be drawn towards the **NASDAQ100 and SP500** synergy. Both of these indices belong to the American market and share numerous characteristics. They encompass American companies traded within the same scenario, which is what makes them a perfect fit for our case.
In this heatmap, they exhibit a vibrant green color, indicative of a high degree of cointegration, or, in simpler terms, a very low probability of not being cointegrated. They demonstrate low p-values suggesting their strength as candidates.

 > 💡 As we can see, this pair of indexes is not the best candidate according to our ADF tests. However, we chose it because the tick data for their prices is publicly available. We used TickStory to obtain the data.

![Prices](https://github.com/hablapps/pairstrading/blob/5-Post/resources/Prices%20gif.gif?raw=true)

> The graphs illustrate the concept of cointegration between two indexes. The top two graphs show the prices of SP500 (left) and NASDAQ100 (right) over the same time period. We can observe that the price movements of these two indices follow similar patterns, suggesting some level of cointegration.

Reviewer: So, what are `USA500` and `USATECH` exactly? What is the connection with these names?

Reviewer 2: The official names are S&P 500 and NASDAQ-100. As far as I know, USA500 and USATECH are synonyms used by some brokers or financial platforms to refer to these indexes

Reviewer: Could we (very briefly) clarify it in the text?

> The bottom graph displays the prices of both indices together, providing a clearer comparison. The blue line represents SP500, and NASDAQ100 is represented by the red line. The close alignment of their price movements indicates that they are cointegrated to some extent. This means that, despite short-term deviations, the indices tend to move together in the long run, maintaining a stable relationship.

Reviewer: Is it ok to talk about "long run" when we are showing just 15 seconds in the last diagram?

```q
system "l pykx.q"
```

Next, we need to import the **statsmodels library**, a prominent tool in Python for statistical modeling and hypothesis testing. It equips analysts with a robust toolkit for regression, time series, and multivariate analysis. Specifically, within the **statsmodels** package, the **statsmodels.tsa.stattools** module features **the Augmented Dickey-Fuller (ADF) test**.

```q
coint:.pykx.import[`statsmodels.tsa.stattools]`:coint
```
For our study, we retrieved data for the different indexes using the [Yahoo Finance API](https://pypi.org/project/yfinance/) and stored them in the `data/stocks/` directory, where we'll find one csv file for each index. Additionally, for simplicity, we only use the closing prices (float), but the API also provides other typical values such as high, low, and open prices.

> We declare the function **read_stock** to read the closing data of a given index.

Reviewer: ```suggestion
We declare the function `rs` (_read stock_) to read the closing data of a given index.

```

> This function uses `0:` to read the files, which takes the delimiter and the schema. In this case, we only want to read the closing price column as a float. Additionally, since the data does not include any reference to the index being read, we need to make a small adjustment to our table to add the index associated with each price. Then, we apply this function to `each` of the indexes from which we want to read the data, concatenate (`raze`) all the data into a table, and finally group (`xgroup`) by index. We declare a variable, **syms**, as a list of symbols, representing each of the indexes we want to check for cointegration.

Reviewer: I would move `Then, we apply...` to another paragraph, so we can split the snippet below, one with `rs` and another one with `syms` and `t`.

Reviewer 2: done

In this heatmap, they exhibit a vibrant green color, indicative of a high degree of cointegration, or, in simpler terms, a very low probability of not being cointegrated. They demonstrate low p-values suggesting their strength as candidates.

 > 💡 As we can see, this pair of indexes is not the best candidate according to our ADF tests. However, we chose it because the tick data for their prices is publicly available. We used TickStory to obtain the data.
![Prices](https://github.com/hablapps/pairstrading/blob/5-Post/resources/Prices%20gif.gif?raw=true)

The graphs illustrate the concept of cointegration between two indexes. The top two graphs show the prices of SP500 (left) and NASDAQ100 (right) over the same time period. We can observe that the price movements of these two indices follow similar patterns, suggesting some level of cointegration.

> The bottom graph displays the prices of both indices together, providing a clearer comparison. The blue line represents SP500, and NASDAQ100 is represented by the red line. The close alignment of their price movements indicates that they are cointegrated to some extent. This means that, despite short-term deviations, the indices tend to move together as time goes on, maintaining a stable relationship.

Reviewer: Not sure if I already mention this in a previous review. What do we consider short-term here? The graphic we are showing is just 15 seconds or so.

Reviewer 2: We've added an image that we think clarifies everything here.

In this case, we are using a [KX Dashboard](https://code.kx.com/dashboards/) to plot our data. We stream this data in one process to our local dashboard, which listens to that process and accesses the data to render visualizations.

## Cointegration, then what?

Let's recap our progress:

1. We've acknowledged the randomness of the market, yet discovered that certain assets **exhibit similar movements**.

2. By employing the **cointegration method and the ADF test**, we pinpointed a promising pair of assets for our analysis: NASDAQ100 and SP500.

> Now we're faced with a crucial question: **"What do I do with these assets?"**

Reviewer: I would be more concrete with the question. Perhaps something like:
_How can I benefit from this knowledge?_

Reviewer 2: done

> # A Match Made in Trading: Step by Step Pairs Trading Guide

Reviewer 3: ```suggestion
# A Match Made in Trading: Step-by-Step Pairs Trading Guide

```

> Q/kdb+ stands out as **a powerful tool in finance**, renowned for its ability to handle vast volumes of real-time data amidst the relentless dynamics of the market. In this article, we embark on an insightful exploration of pair trading and its implementation in Q, offering a comprehensive guide to one of the most popular strategies in the trading world.

Reviewer: Is it 'pair trading' or 'pairs trading'? Stick to one of them.

> Our objective is to **provide a deep understanding of the intricacies of pair trading**, bridging the gap between theory and practice. Through a blend of theoretical insights and practical examples, we aim to equip you with the knowledge and skills necessary to navigate every aspect of this financial modeling strategy.

Reviewer 3: ```suggestion
Our objective is to **provide a deep understanding of the intricacies of Pairs Trading**, bridging the gap between theory and practice. Through a blend of theoretical insights and practical examples, we aim to equip you with the knowledge and skills necessary to navigate every aspect of this financial modelling strategy.

```

Whether taking a technical or quantitative approach, these insights will provide valuable foundations for constructing this algorithm effectively.

## In the untamed realm of the market.

The market has often been described as a **stochastic** (a term which essentially means random) **process** where prices fluctuate irregularly. However, amidst this apparent randomness, we observe that **certain assets move in tandem** due to their inherent relationships. 

For instance, it's logical to expect that if the prices of petrol rise, the prices of cars should also rise. This is because auto companies rely on petrol for their operations, indicating an interconnectedness between the two. Over the long run, they tend to follow similar trends, **reflecting their underlying relationship**.

Is this described mathematically? **Yes**:

The concept we're referring to is **cointegration** (although there are other methods, we'll focus on this one).

> > 💡 Which should not be confused with correlation; cointegration is a statistical property of two time series, indicating a long-term relationship between them despite short-term fluctuations. Cointegrated series move together over time, sharing a common stochastic drift. On the other hand, correlation measures the strength and direction of the linear relationship between two variables at a specific point in time. While correlation captures the degree of association between variables, cointegration reflects a deeper, underlying relationship that persists over time.

Reviewer 3: ```suggestion
> 💡 Which should not be confused with correlation; cointegration is a statistical property of two-time series, indicating a long-term relationship between them despite short-term fluctuations. Cointegrated series move together over time, sharing a common stochastic drift. On the other hand, correlation measures the strength and direction of the linear relationship between two variables at a specific point in time. While correlation captures the degree of association between variables, cointegration reflects a deeper, underlying relationship that persists over time.

```

Hence, we're interested in **cointegrated assets**, which are assets that exhibit the following characteristics:

1. They have a similar trend, meaning the difference between both assets maintains a constant mean, and this difference fluctuates around that same mean.

2. This inherent relationship persists in the long run, meaning that our series is not dependent on time.

## A pair in the hand is worth two in the bush.

> ### ADF testing

Reviewer 3: If you are putting dots `.` at the end of titles, put them in all of them. If you are not, get rid of all of them. My personal suggestion is the latter.

Imagine **we selected 13 world indexes** and aimed to assess whether they are **cointegrated or not**. In this scenario, a crucial tool at our disposal is the Augmented Dickey-Fuller (ADF) test.

> > 💡 Please note that for simplicity in this code, we will be using pykx. This is necessary as we need to import our ADF test function and plot a heatmap of our results.

Reviewer 3: ```suggestion
> 💡 Please note that for simplicity in this code, we will be using PyKX. This is necessary as we need to import our ADF test function and plot a heatmap of our results.

```

> > `system "l pykx.q"`

Reviewer: This lacks context. Is this a q repl? Why placing it as a quote?

> Next, we import the statsmodels.tsa.stattools library and define a custom cointegration function that returns a dictionary with the ADF test results given 2 assets. **We then apply our cointegration function to the Cartesian product of every pair of assets** to get a matrix with every p-value. 

Reviewer: I guess that the library is `statmodels` while `statsmodels.tsa.stattools` is a module? Clarify that this is a python library and briefly describe its objective.

Reviewer: Perhaps you'll explain it later, but at this point I don't understand why you need to return a dictionary with the results of the test, if it just compares two assets. What are the contents of the dictionary? Maybe it would help to briefly describe what is ADF about.

Author: We only need the p-values in our case, but for the sake of generalization we provide the user with every result. And then its his responsability to take whatever he needs (in our case those p-values)

Author: The ADF functionality and porpuse is described and taken care of.

Reviewer 2: done

> Next, we import the statsmodels.tsa.stattools library and define a custom cointegration function that returns a dictionary with the ADF test results given 2 assets. **We then apply our cointegration function to the Cartesian product of every pair of assets** to get a matrix with every p-value. 

Reviewer 3: ```suggestion
Next, we import the `statsmodels.tsa.stattools` library and define a custom cointegration function that returns a dictionary with the ADF test results given 2 assets. **We then apply our cointegration function to the Cartesian product of every pair of assets** to get a matrix with every p-value. 

```

---

> ![cointegration](https://github.com/hablapps/pairstrading/blob/5-Post/resources/CointAlGif.gif?raw=true)

Reviewer: Just a personal opinion: gifs are all right when they add value. In this particular case, it just make it harder to read, since every 5 seconds the diagram disappears.

Reviewer 3: I second this

Author: I will change it

---

```q
// We import the cointegration function from statsmodels library in python
coint:.pykx.import[`statsmodels.tsa.stattools]`:coint; 

> // Recieves 2 symbols and returns a dictionary

Reviewer 3: ```suggestion
// Receives 2 symbols and returns a dictionary

```

fCoint: {p1:0f^(exec close from superTab where sym=x); // AssetX
        p2: 0f^(exec close from superTab where sym=y); // AssetY
        r: 0f^coint[p1;p2]`; // Coint results

>         `pair`score`pvalue`porcentages!(enlist (x,y);r[0];r[1];enlist r[2])}; // Get every result into a dictionary

Reviewer: Several comments on `fCoint`:

- What is `superTab`? You don't mention it and it's crucial.
- There's a typo on `porcentages`.
- In general, this function is mixing responsibilities: it queries the data, calculates the cointegration and formats the result. I think it could be convenient to refactor it, so it takes the close prices as input as well. I'd like to see what are the exact results from `coint`, since I have the feeling that the last line can be improved as well.

Author: 2 first bullet points are taken care of.

Last bullet point Im working on it.

* I added a table to illustrate the contents of the coint results. 
* It is true that it should take price values as inputs.

// We extract every distinct symbol
symList: exec distinct sym from superTab;

> crossedList: symList cross symList

Reviewer: I think that `crossedList` and `symList` also need to be polished, we could be more 'q' idiomatic.

Author: I am not sure if I understand what you mean by this. Should I just join this 2 lines together? and leave it with something like:

`crossedList: x cross x: exec distinct sym from superTab`

> matrix: fCoint .' crossedList;

Reviewer: Is it ok to keep pairs with repeated assets?

Author: Yes, cointegration is 2 directional, meaning that cointegration is different between A-B and B-A, so depends on the direction you are taking you will focus on one set of pairs or the opposite. In order to keep it generic I think this is ok.

```

### Code explanation

> 1. The `coint` function **is imported from statsmodels**, thanks to the pykx library, which provides a cointegration tool in Python. This variable transforms every input data into its Python equivalent, facilitating the calculation of cointegration results.

Reviewer: First off, why do you need PyKX? Because (I guess) `coint` is not available in the standard q libraries. I assume that implementing it from scratch might take a lot of time and would be prone to errors. That's why you need PyKX, since this library enables the integration between Q and Python, son you can rely on `statsmodels`, containing a proven cointegration function that you can rely on. I think that this message is necessary to clarify what is going on here. I won't go into the technical aspects like "transform input data into its Python equivalent", keep it simple.

> 1. The `coint` function **is imported from statsmodels**, thanks to the pykx library, which provides a cointegration tool in Python. This variable transforms every input data into its Python equivalent, facilitating the calculation of cointegration results.

Reviewer 3: ```suggestion
1. The `coint` function **is imported from statsmodels**, thanks to the PyKX library, which provides a cointegration tool in Python. This variable transforms every input data into its Python equivalent, facilitating the calculation of cointegration results.

```

2. We then proceed to **create our custom cointegration function** called `fCoint`, which retrieves both asset prices from our table based on the input symbols. Subsequently, it utilizes the previously imported coint function to obtain the results and inserts them into a dictionary. (`symbol1symbol2...!(;v1;v2;...)`)

> 3. We use QSQL to **extract every distinct symbol** from our table with the keyword [`exec`](https://code.kx.com/q/ref/exec/).

Reviewer 3: ```suggestion
3. We use qSQL to **extract every distinct symbol** from our table with the keyword [`exec`](https://code.kx.com/q/ref/exec/).

```

> > 💡[QSQL](https://code.kx.com/q/basics/qsql/) is a set of feautures that permits users to perform SQL-operations with a very similar syntax on tables in Q/kdb+. Which makes it pretty accesible for newcomers to this language.

Reviewer 3: ```suggestion
> 💡[qSQL](https://code.kx.com/q/basics/qsql/) is a set of features that permits users to perform SQL operations with a very similar syntax on tables in Q/kdb+. Which makes it pretty accessible for newcomers to this language.

```

4. We utilize the `cross` operator, which conducts a [Cartesian product](https://en.wikipedia.org/wiki/Cartesian_product) on every symbol, **generating all possible pair combinations**.

> 5. Finally, **we apply our function from step 2 to our list of every pair of symbols from step 4**. We can achive this by using the [apply operator](https://code.kx.com/q/ref/apply/) (`.`) follow by an [`each`](https://code.kx.com/q/ref/maps/#each) (`'`). So that we apply our function (`fCoint .' `) to each and every pair (`crossedList`)

Reviewer 3: ```suggestion
5. Finally, **we apply our function from step 2 to our list of every pair of symbols from step 4**. We can achieve this by using the [apply operator](https://code.kx.com/q/ref/apply/) (`.`) followed by an [`each`](https://code.kx.com/q/ref/maps/#each) (`'`). So that we apply our function (`fCoint .' `) to each pair (`crossedList`)

```

---

> Now, with our matrix in hand, we can plot it and **visually identify** which asset is more favorable. Given our following assets:

Reviewer 3: ```suggestion
Now, with our matrix in hand, we can plot it and **visually identify** which asset is more favourable. Given our following assets:

```

|   SP500   | NASDAQ100 |   BFX    |   FCHI   |  GDAXI   |   HSI    |   KS11   |   MXX    |   N100   |   N225   |   NYA    |   RUT    |  STOXX   |
|:---------:|:---------:|:--------:|:--------:|:--------:|:--------:|:--------:|:--------:|:--------:|:--------:|:--------:|:--------:|:--------:|
|   USA     |   USA     | Belgium  |  France  | Germany  | Hong Kong|S.Korea   |  Mexico  |  Europe  |  Japan   |   USA    |   USA    |  Europe  |

![ADF heatmap](https://github.com/hablapps/pairstrading/blob/5-Post/resources/ADFgif.gif?raw=true)

> As we can observe, there are several cointegrated indices, but our attention will be drawn towards the **NASDAQ100 and SP500** synergy. Both of these indices belong to the American market and share numerous characteristics. They encompass American companies traded within the same scenario, what makes them a perfect fit for our case.

Reviewer: Explain the plot, what does it mean to have a `0` (green) or `1` (value)? How is this connected with the previous code: score, pvalue, percentage, etc.?

> As we can observe, there are several cointegrated indices, but our attention will be drawn towards the **NASDAQ100 and SP500** synergy. Both of these indices belong to the American market and share numerous characteristics. They encompass American companies traded within the same scenario, what makes them a perfect fit for our case.

Reviewer 3: ```suggestion
As we can observe, there are several cointegrated indices, but our attention will be drawn towards the **NASDAQ100 and SP500** synergy. Both of these indices belong to the American market and share numerous characteristics. They encompass American companies traded within the same scenario, which is what makes them a perfect fit for our case.

```

> ![Prices](https://github.com/hablapps/pairstrading/blob/5-Post/resources/Prices%20gif.gif?raw=true)

Reviewer: Again, briefly explain these graphs.

## Cointegration, then what?

Let's recap our progress:

1. We've acknowledged the randomness of the market, yet discovered that certain assets **exhibit similar movements**.

2. By employing the **cointegration method and the ADF test**, we pinpointed a promising pair of assets for our analysis: NASDAQ100 and SP500.

> 3. These two assets exhibit similar movements and **tend to gravitate around a shared mean**.

Reviewer: Again, I think that it's necessary to explain the previous graphs since you haven't talked about shared means for the pinpointed pair.

Author: I will

Now we're faced with a crucial question: **"What do I do with these assets?"**

As mentioned earlier, the market is inherently random and doesn't always behave predictably. While NASDAQ100 and SP500 often follow similar trends, their individual values **can sometimes diverge significantly**. For instance, NASDAQ100 may rise while SP500 falls, or vice versa. 

However, this presents **an opportunity for profit** because we know that these assets tend to revert to their shared mean over time. If one asset is **overpriced** and likely to decrease, we may consider **selling it** (going short). Conversely, if an asset is **underpriced** and expected to increase, we may consider **buying it** (going long). And that is what we call Pairs Trading.

> 💡 This strategy possesses intriguing financial characteristics: our **profitability remains unaffected by the broader market trends**, as our focus lies solely on the disparity between the two assets. It's about relative movements rather than absolute ones; we're indifferent to whether prices are rising or falling. This quality defines it as a **neutral market strategy**.

## Spreading spreads

> Indeed, just subtracting the prices of two assets, as in: $priceY−priceX$ may not provide a clear understanding of their relationship. Let's illustrate this with an example:

Reviewer: In my view, this section starts abruptly. What are we connecting with this *Indeed*? At this point, I'm not sure what this section is about.

> Indeed, just subtracting the prices of two assets, as in: $priceY−priceX$ may not provide a clear understanding of their relationship. Let's illustrate this with an example:

Reviewer 3: ```suggestion
Indeed, just subtracting the prices of two assets, as in $priceY−priceX$ may not provide a clear understanding of their relationship. Let's illustrate this with an example:

```

Consider the following series:

---

```q
priceX: 5 10 7 4 8

priceY: 23 30 25 30 35

> spreads: priceY - priceX // = 18 20 18 26 27

Reviewer: Why not using a REPL notation so you don't need to show input as a comment?

Author: I am not familiar with REPL notation. This is written as code because Christian and I wanted to show Q array capabilities and we wanted also to be somewhat technical about this post showing some simple code

```

> 💡 As you can see and verify through Q/kdb+ array properties, spreads can be calculated by simply computing the difference between two vectors without the need for special functions or loops.

---

〽 **These spread values don't offer much insight** into the relationship between the two assets. Are both assets increasing? Are they moving in opposite directions? It's unclear from these numbers alone.

> Let's consider **using logarithms**, as they possess favorable properties for our pricing model. They inherently prevent negative values and tend to approach zero:

Reviewer 3: ```suggestion
Let's consider **using logarithms**, as they possess favourable properties for our pricing model. They inherently prevent negative values and tend to approach zero:

```

--- 

```q
log priceX = 1.609438 2.302585 1.94591 1.386294 2.079442

log priceY = 3.135494 3.401197 3.218876 3.401197 3.555348

spreads: log[priceY] - log priceX // = 1.526056 1.098612 1.272966 2.014903 1.475907
```

---

We're making progress, as we observe **numbers now fluctuating within much smaller ranges**. However, we're still missing a clear understanding of the underlying relationship. While we've normalized the data using logarithms, we now need to align their discrepancies to a single asset. 

Since both assets are related, **we can leverage linear regression** to our advantage. This enables us to simplify our spreads effectively. So, we'll conduct a basic linear regression analysis using historical data to discern the disparity between them:

--- 

$spread = log(priceY) - (beta * log(priceX)+alpha)$

```q
historical_data_priceX: 7 10 6 5 8

historical_data_priceY: 23 25 16 20 15

beta: betaF[historical_data_priceX;historical_data_priceY] // = 0.2679227

alpha: alphaF[historical_data_priceX;historical_data_priceY] // = 2.444817

> ```

Reviewer: If you want to map a formula to code, it's better if you put them together, and show that there's actually a correspondence among them. Here, you are showing the formula for spread, but the implementation for `beta` and `alpha`. Later, in a different subsection, you show the formulas for `beta` and `alpha`. And finally you show the implementation for `spread`. This is all messy and difficult to follow. In addition to that, I miss the implementation of `betaF` and `alphaF`. Could you please reorder it and make it self-content to improve the text flow?

Author: Should I add the `betaF` and `alphaF` to this post then?

My idea was:

1. Showing that we need to calculate beta and alpha for some data
2. Briefly explain these formulas 

You prefer this I am understanding:

1. Explain formulas
2. Show its implementation
3. Use them in our spreads calculation

I will change it

### Code Explanation

As you can observe, we utilize the `betaF` and `alphaF` functions to compute the beta and alpha coefficients for a given dataset. Specifically, these functions perform linear regression to estimate alpha and beta based on historical values.

The implementation of these functions is based on the following formulas:

$\beta = \frac{{(n \cdot \sum(x \cdot y)) - (\sum x \cdot \sum y)}}{{(n \cdot \sum(x^2)) - (\sum x)^2}}$

$\alpha = \text{Mean}(y) - \beta \cdot \text{Mean}(x)$

---

We've already calculated the alpha and beta using the logarithmic values of our historical data (since we don't have prior knowledge of the real-time price values for priceX and priceY). Now, all that remains is to apply the previous formula to derive our spreads:

---

```q
spreads: historical_data_priceY - ((historical_data_priceX*beta)+alpha) // = -0.1493929 0.0451223 -0.08835117 0.0451223 0.1579725
```

> 🖥️ In Q/kdb+, operand priority is strictly from right to left, without any precedence rules except those involving parentheses. Therefore, it's crucial to exercise caution when writing Q code to ensure accurate results.

---

> And this precisely meets our objective—a **comprehensive method for representing relative changes between both assets**. As we can deduce, our mean is now 0 because our assets are normalized, cointegrated and on the same scale. Therefore, ideally, the differential between their prices should be 0. Consequently, when our spread is below 0, we infer that asset X is overpriced, whereas if it's above 0, then asset Y is overpriced.

Reviewer 3: ```suggestion
This precisely meets our objective—a **comprehensive method for representing relative changes between both assets**. As we can deduce, our mean is now 0 because our assets are normalized, cointegrated and on the same scale. Therefore, ideally, the differential between their prices should be 0. Consequently, when our spread is below 0, we infer that asset X is overpriced, whereas if it's above 0, then asset Y is overpriced.

```

![Spreads](https://github.com/hablapps/pairstrading/blob/5-Post/Spreads.gif?raw=true)

## Two steps forward, one step back.

> Before proceeding to plot the NASDAQ100-SP500 spreads, we need to first plan our algorithm. In this post, we intend to create **a real-time scenario** for Pairs trading, so careful planning is essential for our sake.

Reviewer 3: ```suggestion
Before proceeding to plot the NASDAQ100-SP500 spreads, we need to first plan our algorithm. In this post, we intend to create **a real time scenario** for Pairs Trading, so careful planning is essential for our sake.

```

Decomposing our steps, let's start from the very beginning:

1. Data management part

        1.1. Read data

        1.2. Filter data

> 2. Outside the loop

Reviewer: What loop? Is this intended to be some kind of pseudocode? I find it confusing.

        2.1. Linear regression

        2.2. Initialize values 

3. Inside the loop

        3.1. Calculate spreads

        3.2. Write on buffers

As previously mentioned, historical data is crucial for generating accurate spreads. We need to calculate each spread in real-time **using precomputed alpha and beta values** derived from both prices. Therefore, once we have obtained our historical values, we can proceed with calculating our linear regression.

---

```q
// Fix data and take log(prices) -> Simulated data
priceX: 0!1_(update delta:0f^deltas dateTime from 
        distinct select distinct dateTime, log bid, log ask from
        (update dateTime:"P"$@[;19;:;"."] each dateTime from 
        tab1) where not null bid);

priceY: 0!1_(update delta:0f^deltas dateTime from 
        distinct select distinct dateTime, log bid, log ask from 
        (update dateTime:"P"$@[;19;:;"."] each dateTime from 

>         tab2) where not null bid);

Reviewer: These definitions are essentially the same, up to the table they read from. Modularise it. By the way, what is `tab1` and `tab2`? What is going on with the hardcoded `19` value?

Author: I will take a look at that

// Read historical data
historial_tab2: 1_ flip `open`high`low`close`adjClose`vol!("FFFFFF";",") 0: `:/data/stocks/NASDAQ100_hist.csv;

> historial_tab1: 1_ flip `open`high`low`close`adjClose`vol!("FFFFFF";",") 0: `:data/stocks/SP500_hist.csv;

Reviewer: Again, both definitions are pretty much the same, modularise it.

Reviewer: What is the point of reading so many columns if you only use `close`?

Author: In my opinion modulerising it adds a layer of abstraction, and thus complexity. Wouldn't it be better to duplicate these definitions so it is easier to understand?

Reading every column is because Pairs trading can be performed by focusing on different prices; in this case we use close price, but we could also use open price or whatever we want

```

---

## Linear regression

Now, armed with logarithms, we can replicate the process from our previous example and calculate both alpha and beta values. To do this, we'll start by taking the historical data for NASDAQ100 and SP500 and applying our linear regression function to derive the desired values.

---

```q
// Calculate alpha and beta from historical values
beta_lr: betaF[px:-100#log historial_tab1`close;py:-100#log historial_tab2`close]; // we only take most recent 100 values 

alpha_lr: alphaF[px;py];
```

---

Now we just calculate our spreads as we did before like:

---

```q

> spread: priceY[.streamPair.i][`bid] - ((priceX[.streamPair.i][`bid] * beta_lr)+alpha_lr);

Reviewer: What is `.streamPair.i`?

```

---

> > 💡 You may notice that we retrieve bid price data from our price stream using an index (`.streamPair.i`). This is because we are simulating the arrival of these records dynamically, based on some delta time.

Reviewer: Ok, now `.streamPair.i` is explained, but to be honest, I don't get this explanation.

This approach will provide us with:

![SpreadsD](https://github.com/hablapps/pairstrading/blob/5-Post/resources/Spreads%20gif.gif?raw=true)

And there we have it! **A perfectly plotted spread series in real-time**, ready to be utilized for further analysis and exploitation.

> In this case we are using a [KX Dashboard](https://code.kx.com/dashboards/) to plot our data. We stream this data in one process to our local dashboard, which listens to that process and accesses the data to render visualizations.

Reviewer 3: ```suggestion
In this case, we are using a [KX Dashboard](https://code.kx.com/dashboards/) to plot our data. We stream this data in one process to our local dashboard, which listens to that process and accesses the data to render visualizations.

```

> In this case we are using a [KX Dashboard](https://code.kx.com/dashboards/) to plot our data. We stream this data in one process to our local dashboard, which listens to that process and accesses the data to render visualizations.

Reviewer: Dashboards have already been used above. Move this link to that place.

> ## What's left to start making profit? 

Reviewer 3: ```suggestion
## What's left to start making a profit? 

```

> Finally, once we have our spreads accurately calculated and observe how our data is being updated second by second, we can **execute buy and sell orders when spreads discrepancies occur** based on some signal windows. Those windows, however, will be explored in greater depth in our next post about the Kalman Filter and its application in Pairs Trading.

Reviewer 3: ```suggestion
Finally, once we have our spreads accurately calculated and observe how our data is being updated second by second, we can **execute buy and sell orders when spread discrepancies occur** based on some signal windows. Those windows, however, will be explored in greater depth in a follow up post about the Kalman Filter and its application in Pairs Trading.

```

> 💡 Signal windows play a pivotal role in implementing Pairs Trading strategies. They serve as indicators for determining when to execute buy and sell actions, acting as arbitrary thresholds that guide our algorithm's decision-making process. These windows are derived from the variance of our data, representing a static variance assumption due to our consideration of a time-independent cointegrated series. However, we'll delve deeper into this topic in a subsequent post that will expand the scope of the current discussion as we previously mentioned.

For now, it's crucial to clarify **our spread formulation and understand what it represents**. With this knowledge, we can identify instances where one asset is overpriced while the other is underpriced.

> One might argue that our calculations are heavily influenced by past data, and that we rely too much on historical changes that **may not accurately reflect the present reality**. This is indeed a **valid concern**. To address this issue, we can utilize **the Kalman Filter**, a mathematical method for filtering noise and predicting states in a dynamic system. But we'll delve into the Kalman Filter in our upcoming posts as previously mentioned.

Reviewer 3: ```suggestion
One might argue that our calculations are heavily influenced by past data and that we rely too much on historical changes that **may not accurately reflect the present reality**. This is indeed a **valid concern**. To address this issue, we can utilize **the Kalman Filter**, a mathematical method for filtering noise and predicting states in a dynamic system. But we'll delve into the Kalman Filter in our upcoming posts as previously mentioned.

```

> One might argue that our calculations are heavily influenced by past data, and that we rely too much on historical changes that **may not accurately reflect the present reality**. This is indeed a **valid concern**. To address this issue, we can utilize **the Kalman Filter**, a mathematical method for filtering noise and predicting states in a dynamic system. But we'll delve into the Kalman Filter in our upcoming posts as previously mentioned.

Reviewer: This is the second reference to a future Kalman filter post in this section. I think there's some redundancy with it.

Author: This is true but I am figuring out how to put it. Because next post will explain two different things regarding pairs trading:

1. Signal windows
2.  Kalman filter

Additionally, even though we fit our model with historical data, we could implement **a rolling window approach** where the linear regression is continuously updated. This would ensure that our model remains responsive to changes in the underlying data over time.

## The End

In conclusion, this post aims to provide a comprehensive overview that explains Pairs Trading from beginning to end, covering every aspect that may be relevant to understanding this topic and its intricacies in Q/kdb+.

Recapping, we have covered:

1. An introductory overview of the market as a whole.
2. An examination of cointegrated assets within the market.
3. Multiple Augmented Dickey-Fuller (ADF) tests on real assets.
4. A presentation of the pairs trading strategy itself.

> 5. A clear and guided explanation of spreads calculation, interpretation, and implementation in Q/kdb+.

Reviewer 3: ```suggestion
5. A clear and guided explanation of spread calculation, interpretation, and implementation in Q/kdb+.

```

6. Additional knowledge necessary to master the strategy.

> We aimed to demonstrate the capabilities of Q/kdb+ and its potential in a simplified manner that anyone can implement, particularly in the context of a widely used financial strategy. By doing so, we hope to make complex concepts more accessible and empower individuals to leverage these powerful tools in their own endeavors.

Reviewer 3: ```suggestion
We aimed to demonstrate the capabilities of Q/kdb+ and its potential in a simplified manner that anyone can implement, particularly in the context of a widely used financial strategy. By doing so, we hope to make complex concepts more accessible and empower individuals to leverage these powerful tools in their own endeavours.

```

> We hope you found this information valuable and gained a good understanding of this financial tactic from both technical and economical perspectives. If you have any questions or need further clarification, don't hesitate to reach out. 

Reviewer 3: ```suggestion
We hope you found this information valuable and gained a good understanding of this financial tactic from both technical and economic perspectives. If you have any questions or need further clarification, don't hesitate to reach out. 

```

Be sure to stay tuned for more posts and updates on this blog to deepen your knowledge even further. 

> Special thanks to [...] for [...]

Reviewer 3: This seems to be unfinished...

Reviewer: Complete the "special thanks", don't forget Javier.

## References and Documentation

> For the technical implementation we relied on:

Reviewer 3: ```suggestion
For the technical implementation, we relied on:

```

* Kx Documentation: https://code.kx.com/q/ref/
* Q for mortals: https://code.kx.com/q4m3/
* Pykx Documentation: https://code.kx.com/pykx/2.4/index.html
* statsmodels Documentation: https://www.statsmodels.org/dev/generated/statsmodels.tsa.stattools.coint.html

> For the financial implementation we used:

Reviewer 3: ```suggestion
For the financial implementation, we used:

```
